src/workerd/api/node/zlib-util.c++

```diff
@@ -3,9 +3,159 @@
 //     https://opensource.org/licenses/Apache-2.0
 #include "zlib-util.h"
 
+#include <workerd/jsg/exception.h>
+
+#include <zlib.h>
+
+#include <iostream>
+
 namespace workerd::api::node {
 uint32_t ZlibUtil::crc32Sync(kj::Array<kj::byte> data, uint32_t value) {
   // Note: Bytef is defined in zlib.h
   return crc32(value, reinterpret_cast<const Bytef*>(data.begin()), data.size());
 }
+
+// Simple wrapper of kj::Vector for use as a buffer
+class GrowableBuffer: public kj::Vector<kj::byte> {
+  public:
+  kj::byte* writePos() {
+    // Ensure there's a backing array allocated
+    reserve(1);
+
+    return end();
+  }
+
+  size_t available() const {
+    return capacity() - size();
+  }
+
+  void returnUnused(size_t unused) {
+    resize(capacity() - unused);
+  }
+
+  void growBy(size_t extra) {
+    reserve(size() + extra);
+  }
+};
+
+
+template <typename Impl> class ZlibSyncContext {
+public:
+  ZlibSyncContext(kj::ArrayPtr<kj::byte> data, const ZlibUtil::Options& options): options(options) {
+    stream.next_in = data.begin();
+    stream.avail_in = data.size();
+    auto status = impl().init();
+    JSG_REQUIRE(status == Z_OK, Error, "Failed to initialize zlib");
+  }
+
+  kj::Array<kj::byte> processBuffer() {
+    int status = Z_OK;
+
+    do {
+      result.growBy(ZLIB_OUT_BUF_SIZE);
```

Comment:
This should incorporate a check to make sure the `result` does not grow beyond a reasonable limit (I'd say max `128 * 1024`).